File: PersianCalendar/src/main/java/net/androgames/level/orientation/OrientationProvider.java
package net.androgames.level.orientation;

import android.app.Activity;
import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.view.Surface;

import net.androgames.level.view.LevelView;

import java.util.List;

/*
 *  This file is part of Level (an Android Bubble Level).
 *  <https://github.com/avianey/Level>
 *
 *  Copyright (C) 2014 Antoine Vianey
 *
 *  Level is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Level is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Level. If not, see <http://www.gnu.org/licenses/>
 */
public final class OrientationProvider implements SensorEventListener {

    private static final int MIN_VALUES = 20;
    /**
     * Rotation Matrix
     */
    private final float[] MAG = new float[]{1f, 1f, 1f};
    private final float[] I = new float[16];
    private final float[] R = new float[16];
    private final float[] outR = new float[16];
    private final float[] LOC = new float[3];
    private Sensor sensor;
    private SensorManager sensorManager;
    /**
     * indicates whether or not Accelerometer Sensor is running
     */
    private boolean running = false;
    /**
     * Orientation
     */
    private float pitch;
    private float roll;
    private float balance;
    private float minStep = 360;
    private float refValues = 0;
    private int displayOrientation;
    private LevelView view;

    public OrientationProvider(Activity activity, LevelView view) {
        this.view = view;
        this.displayOrientation = activity.getWindowManager().getDefaultDisplay().getRotation();
        this.sensorManager = (SensorManager) activity.getSystemService(Context.SENSOR_SERVICE);
        if (sensorManager == null) return;

        List<Sensor> sensors = sensorManager.getSensorList(Sensor.TYPE_ACCELEROMETER);
        if (sensors.size() == 0) return;

        this.sensor = sensors.get(0);
    }

    /**
     * Returns true if the manager is listening to orientation changes
     */
    public boolean isListening() {
        return running;
    }

    /**
     * Unregisters listeners
     */
    public void stopListening() {
        running = false;
        try {
            if (sensorManager != null) {
                sensorManager.unregisterListener(this);
            }
        } catch (Exception ignore) {
        }
    }

    /**
     * Registers a listener and start listening
     */
    public void startListening() {
        // register listener and start listening
        if (sensorManager == null || sensor == null) return;
        running = sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_FASTEST);
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }

    @Override
    public void onSensorChanged(SensorEvent event) {

        float oldPitch = pitch;
        float oldRoll = roll;
        float oldBalance = balance;

        SensorManager.getRotationMatrix(R, I, event.values, MAG);

        // compute pitch, roll & balance
        switch (displayOrientation) {
            case Surface.ROTATION_270:
                SensorManager.remapCoordinateSystem(
                        R,
                        SensorManager.AXIS_MINUS_Y,
                        SensorManager.AXIS_X,
                        outR);
                break;
            case Surface.ROTATION_180:
                SensorManager.remapCoordinateSystem(
                        R,
                        SensorManager.AXIS_MINUS_X,
                        SensorManager.AXIS_MINUS_Y,
                        outR);
                break;
            case Surface.ROTATION_90:
                SensorManager.remapCoordinateSystem(
                        R,
                        SensorManager.AXIS_Y,
                        SensorManager.AXIS_MINUS_X,
                        outR);
                break;
            case Surface.ROTATION_0:
            default:
                SensorManager.remapCoordinateSystem(
                        R,
                        SensorManager.AXIS_X,
                        SensorManager.AXIS_Y,
                        outR);
                break;
        }

        SensorManager.getOrientation(outR, LOC);

        // normalize z on ux, uy
        float tmp = (float) Math.sqrt(outR[8] * outR[8] + outR[9] * outR[9]);
        tmp = (tmp == 0 ? 0 : outR[8] / tmp);

        // LOC[0] compass
        pitch = (float) Math.toDegrees(LOC[1]);
        roll = -(float) Math.toDegrees(LOC[2]);
        balance = (float) Math.toDegrees(Math.asin(tmp));

        // calculating minimal sensor step
        if (oldRoll != roll || oldPitch != pitch || oldBalance != balance) {
            if (oldPitch != pitch) {
                minStep = Math.min(minStep, Math.abs(pitch - oldPitch));
            }
            if (oldRoll != roll) {
                minStep = Math.min(minStep, Math.abs(roll - oldRoll));
            }
            if (oldBalance != balance) {
                minStep = Math.min(minStep, Math.abs(balance - oldBalance));
            }
            if (refValues < MIN_VALUES) {
                refValues++;
            }
        }

        Orientation orientation;
        if (pitch < -45 && pitch > -135) {
            // top side up
            orientation = Orientation.TOP;
        } else if (pitch > 45 && pitch < 135) {
            // bottom side up
            orientation = Orientation.BOTTOM;
        } else if (roll > 45) {
            // right side up
            orientation = Orientation.RIGHT;
        } else if (roll < -45) {
            // left side up
            orientation = Orientation.LEFT;
        } else {
            // landing
            orientation = Orientation.LANDING;
        }

        // propagation of the orientation
        view.setOrientation(orientation, pitch, roll, balance);
    }
}

File: PersianCalendar/src/main/java/net/androgames/level/OrientationProvider.kt
package net.androgames.level

import android.app.Activity
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.view.Surface

/*
*  This file is part of Level (an Android Bubble Level).
*  <https://github.com/avianey/Level>
*
*  Copyright (C) 2014 Antoine Vianey
*
*  Level is free software: you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation, either version 3 of the License, or
*  (at your option) any later version.
*
*  Level is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with Level. If not, see <http://www.gnu.org/licenses/>
*/
class OrientationProvider(activity: Activity, private val view: LevelView) : SensorEventListener {
    /**
     * Rotation Matrix
     */
    private val MAG = floatArrayOf(1f, 1f, 1f)
    private val I = FloatArray(16)
    private val R = FloatArray(16)
    private val outR = FloatArray(16)
    private val LOC = FloatArray(3)
    private val sensorManager: SensorManager?
    private val displayOrientation: Int
    private val sensor: Sensor?
    /**
     * Returns true if the manager is listening to orientation changes
     */
    /**
     * indicates whether or not Accelerometer Sensor is running
     */
    var isListening = false
        private set

    /**
     * Orientation
     */
    private var pitch = 0f
    private var roll = 0f
    private var balance = 0f
    private var minStep = 360f
    private var refValues = 0f

    /**
     * Unregisters listeners
     */
    fun stopListening() {
        isListening = false
        try {
            sensorManager?.unregisterListener(this)
        } catch (ignore: Exception) {
        }
    }

    /**
     * Registers a listener and start listening
     */
    fun startListening() {
        // register listener and start listening
        if (sensorManager == null || sensor == null) return
        isListening =
            sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_FASTEST)
        view.invalidate()
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
    override fun onSensorChanged(event: SensorEvent) {
        val oldPitch = pitch
        val oldRoll = roll
        val oldBalance = balance
        SensorManager.getRotationMatrix(R, I, event.values, MAG)
        when (displayOrientation) {
            Surface.ROTATION_270 -> SensorManager.remapCoordinateSystem(
                R,
                SensorManager.AXIS_MINUS_Y,
                SensorManager.AXIS_X,
                outR
            )
            Surface.ROTATION_180 -> SensorManager.remapCoordinateSystem(
                R,
                SensorManager.AXIS_MINUS_X,
                SensorManager.AXIS_MINUS_Y,
                outR
            )
            Surface.ROTATION_90 -> SensorManager.remapCoordinateSystem(
                R,
                SensorManager.AXIS_Y,
                SensorManager.AXIS_MINUS_X,
                outR
            )
            Surface.ROTATION_0 -> SensorManager.remapCoordinateSystem(
                R,
                SensorManager.AXIS_X,
                SensorManager.AXIS_Y,
                outR
            )
            else -> SensorManager.remapCoordinateSystem(
                R,
                SensorManager.AXIS_X,
                SensorManager.AXIS_Y,
                outR
            )
        }
        SensorManager.getOrientation(outR, LOC)

        // normalize z on ux, uy
        var tmp = Math.sqrt((outR[8] * outR[8] + outR[9] * outR[9]).toDouble())
            .toFloat()
        tmp = if (tmp == 0f) 0 else outR[8] / tmp

        // LOC[0] compass
        pitch = Math.toDegrees(LOC[1].toDouble()).toFloat()
        roll = (-Math.toDegrees(LOC[2].toDouble())).toFloat()
        balance = Math.toDegrees(Math.asin(tmp.toDouble())).toFloat()

        // calculating minimal sensor step
        if (oldRoll != roll || oldPitch != pitch || oldBalance != balance) {
            if (oldPitch != pitch) {
                minStep = Math.min(minStep, Math.abs(pitch - oldPitch))
            }
            if (oldRoll != roll) {
                minStep = Math.min(minStep, Math.abs(roll - oldRoll))
            }
            if (oldBalance != balance) {
                minStep = Math.min(minStep, Math.abs(balance - oldBalance))
            }
            if (refValues < MIN_VALUES) {
                refValues++
            }
        }
        val orientation: Orientation
        orientation = if (pitch < -45 && pitch > -135) {
            // top side up
            Orientation.TOP
        } else if (pitch > 45 && pitch < 135) {
            // bottom side up
            Orientation.BOTTOM
        } else if (roll > 45) {
            // right side up
            Orientation.RIGHT
        } else if (roll < -45) {
            // left side up
            Orientation.LEFT
        } else {
            // landing
            Orientation.LANDING
        }

        // propagation of the orientation
        view.setOrientation(orientation, pitch, roll, balance)
    }

    companion object {
        private const val MIN_VALUES = 20
    }

    init {
        displayOrientation = activity.windowManager.defaultDisplay.rotation
        sensorManager = activity.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        if (sensorManager == null) return
        val sensors = sensorManager!!.getSensorList(Sensor.TYPE_ACCELEROMETER)
        if (sensors.size == 0) return
        sensor = sensors[0]
    }
}
